tasks/libs/pipeline_notifications.py

```diff
@@ -7,7 +7,7 @@
 from .types import Test
 
 
-def get_failed_jobs(project_name, pipeline_id):
+def get_failed_jobs(project_name, pipeline_id, include_allow_failures=False):
```

Comment:
Would you mind moving the allow failure logic into a separate PR? It will make review easier

Reply:
Also, I am not sure I understand this
> we don't want to fail unrelated changes due to a new Go version being released (which could happen at any time)

Why can't we pin the Go version instead? The one we use in the builds is pinned and we have to manually update it. Why wouldn't we be able to update the lookup tables then?

Reply:
> Would you mind moving the allow failure logic into a separate PR? It will make review easier

Sure, I created a new PR here: https://github.com/DataDog/datadog-agent/pull/10057

> we don't want to fail unrelated changes due to a new Go version being released (which could happen at any time)

Sorry, I probably explained this poorly. What the rest of this PR does is add a system for running commands on a bunch of Go versions (right now it does 1.15, 1.16, and 1.17, but once 1.18 is released then the code will also download and run that toolchain) in order to try to extract information from binaries that might change depending on the version of the Go compiler/standard library. This information will then be used by a tool within system-probe that is designed to work with *any* arbitrary customer-supplied binary compiled with Go (no matter the version of Go it was compiled with, as long as it's greater than some min supported version). Because of that, we want to be able to quickly update the generated code if a new Go release ends up changing the output of the compiler so that the next agent release will work with binaries compiled by the new version of Go. In this way, pinning the version of Go defeats the purpose of getting this information from the binaries, since we want to know if/how it changes between Go versions.

Reply:
Thanks for the detailed explanation! It makes sense to quickly test new Go versions, and I understand what your proposed solution is.

IMHO, I think it would be simpler to setup a Datadog monitor for new Go versions that notifies your team instead of doing this workaround with 'allow to fail'. We have a similar setup for being notified for new certificate bundles, and can help you do a similar thing for this if you think that approach makes sense

Reply:
I'll admit that I'm not very familiar with Datadog monitors, but it seems to me like using a monitor might not work as well. At least with the way I'm envisioning it, a monitor could somehow get the latest Go version (there's a simple JSON API for this) and then compare it to a hard-coded version in the monitor, and if it's greater, send an alert. This would accomplish our goals of having a reliable method of getting alerted when new Go versions are released that might result in us needing to re-generate the lookup tables, and the alerts would be persistent/can’t easily be ignored. However, it seems like such a monitor would cause a lot more noise than the current proposed solution, since it's pretty common that a new Go release doesn't actually change any of the offsets/information we’re obtaining from the binaries (in which case we can ignore it). With the current proposed solution, we will only receive an alert if a new Go release actually changes the generated lookup tables; otherwise it won't fail/send a notification. Also, it seems like the monitor would require more manual work, since we would have to both check/update the lookup tables *and* update the hard-coded minimum version in the monitor whenever a new Go version is released.

With that in mind:
- How strongly do you prefer another solution over the proposed “allowed-to-fail” semantics?
- Is there a better way of using a monitor that lets us accomplish the same thing as our "allowed-to-fail" changes?
- Are there any better/easier ways of accomplishing our goals (perhaps some form of nightly CI job)?

Reply:
Hey there, let me try to answer these

> How strongly do you prefer another solution over the proposed “allowed-to-fail” semantics?

We discussed this in agent platform and we see the following concerns with the 'allowed-to-fail' approach:
1. Whenever this starts failing because of a new version, it will fail for all pipelines on main, which will be very noisy for your team (you will be notified ~10 times/day until you fix it)
2. If this fails during a merge freeze during a late RC, we would need to do a PR to fix this to avoid the spam. Such a PR can't in principle add a new feature (support a new Go version), so we would have to add a mechanism to turn off the notifications in some way and do PRs mid-freeze just for this
3. If this code suffers a regression and starts failing for all Go versions, we won't have easy visibility into it until things are merged to main

> Is there a better way of using a monitor that lets us accomplish the same thing as our "allowed-to-fail" changes?

If I understand correctly, your concern is that a monitor that just checks new Go versions is going to be noisy since you won't care about most Go versions. I think per (1) above this would be much less noisy than the 'allowed-to-fail' approach. I would recommend you to try this monitor out and iterate from there if it's too noisy.

> Are there any better/easier ways of accomplishing our goals (perhaps some form of nightly CI job)?

If the 'check if there is a new Go version' monitor doesn't work for you, an alternative would be to send a metric after running your test in a Gitlab scheduled pipeline and trigger a monitor with that?

Reply:
> Whenever this starts failing because of a new version

It is worth pointing out that this won't fail on just a new Go version. It would only fail on a new Go version that has *different offsets*.

> I would recommend you to try this monitor out

Can you point us to your certificate bundle monitor as an example to work from? Thanks!